R/pkg/tests/fulltests/test_Serde.R

```diff
@@ -123,6 +123,49 @@ test_that("SerDe of list of lists", {
   expect_equal(x, y)
 })
 
+test_that("SPARK-20007 *apply() can return empty data.frame", {
+  df <- createDataFrame(
+    list(list(1L, 1, "1"), list(2L, 2, "2"), list(3L, 3, "3")),
+    c("a", "b", "c"))
+
+  # dapply
+  schema1 <- structType(structField("a", "integer"), structField("b", "double"),
+                         structField("c", "string"))
+  func1 <- function(x) {
+    y <- x[x[1] > 3, ]
+  }
+
+  df1 <- dapply(df, func1, schema1)
+  result1 <- collect(df1)
+
+  # construct expected1
+  expected1 <- data.frame(matrix(ncol = 3, nrow = 0))
+  names <- c("a", "b", "c")
+  colnames(expected1) <- names
+  for (k in names) expected1[[k]] <- list()
+
+  # The following should be the correct way to generate expected1,
+  # but result1 can not match this in the current fix.
+  # x <- collect(df)
+  # expected1 <- func1(x)
```

Comment:
@falaki In my current (simple) fix, the data.frame result1 has columns of `list()`, instead of the original column types. Ideally, we should achieve this for correctness. I'll need more time to figure out how to improve it. Do you have comments or suggestions for the current direction?